Sources/SecureStorage/SecureStorageDatabaseProvider.swift

```diff
@@ -0,0 +1,147 @@
+//
+//  SecureStorageDatabaseProvider.swift
+//
+//  Copyright © 2023 DuckDuckGo. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+import Common
+import GRDB
+
+public protocol SecureStorageDatabaseProvider {
+
+    var db: DatabaseWriter { get }
+
+}
+
+open class GRDBSecureStorageDatabaseProvider: SecureStorageDatabaseProvider {
+
+    public enum DatabaseWriterType {
+        case queue
+        case pool
+    }
+
+    public let db: DatabaseWriter
+
+    public init(file: URL, key: Data, writerType: DatabaseWriterType = .queue, registerMigrationsHandler: (inout DatabaseMigrator) throws -> Void) throws {
+        do {
+            self.db = try Self.createDatabase(file: file, key: key, writerType: writerType, registerMigrationsHandler: registerMigrationsHandler)
+        } catch SecureStorageDatabaseError.corruptedDatabase {
+            try Self.recreateDatabase(withKey: key, databaseURL: file)
+            self.db = try Self.createDatabase(file: file, key: key, writerType: writerType, registerMigrationsHandler: registerMigrationsHandler)
+        }
```

Comment:
The GRDB database provider now takes care of the backup functionality as a part of its initialization - added a unit test for this also.